File: filter_debug.py
#!/usr/bin/env python3
"""
実際のYahoo API + fxtwitter結果でフィルタ通過/除外を調査するスクリプト
使い方: python3 filter_debug.py [クエリ] [件数=40]
"""
import sys, urllib.request, json, re, concurrent.futures

QUERY = sys.argv[1] if len(sys.argv) > 1 else "イラスト"
RESULTS = int(sys.argv[2]) if len(sys.argv) > 2 else 40

UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
YAHOO_HEADERS = {"User-Agent": UA, "Accept": "application/json", "Referer": "https://search.yahoo.co.jp/realtime/search"}

ART_KEYWORDS_JP = ["イラスト", "絵師", "描き", "描いて", "お絵描", "創作", "漫画家", "fanbox", "skeb", "pixiv", "イラストレーター"]
ART_KEYWORDS_EN = re.compile(r'\b(artist|illustration|illustrator|drawing|fanart|manga|comic)\b|digital\s+art|fan\s+art|pixel\s+art', re.I)
ART_SITES = ["pixiv", "skeb", "fanbox", "deviantart", "artstation", "nijie", "chichi-pui", "aipictors", "patreon"]
NEG_PHRASES_BIO = ["イラスト好き", "イラストまとめ", "イラストメディア", "イラスト収集", "イラスト垢", "絵が好き", "絵を集め", "fanart好き", "fan art好き", "描いてもらった", "描いていただ"]
POS_PHRASES_BIO = ["絵を描", "イラストを描", "描いてます", "描いています", "お絵描きしてます", "絵描きです", "イラスト描いてます", "漫画描いてます"]
NON_ART_BIO = ["公式", "official", "情報発信", "ニュース", "news", "グッズ", "フィギュア", "コスプレ", "cosplay", "ダンス", "dance", "歌", "singer", "vtuber", "配信", "料理", "レシピ", "グルメ", "食べ", "メディア欄", "まとめ", "推し活", "写真", "photo", "photographer", "カメラ", "撮影"]

def is_illustrator(author):
    desc = author.get("description", "") or ""
    loc = author.get("location", "") or ""
    name = author.get("name", "") or ""
    site = (author.get("website", {}) or {}).get("url", "") or ""
    bio_all = desc + " " + loc + " " + name
    bio_lower = bio_all.lower()

    if any(p in bio_all for p in NEG_PHRASES_BIO):
        return False, "NEG_PHRASE: " + next(p for p in NEG_PHRASES_BIO if p in bio_all)

    pos_phrase = any(p in bio_all for p in POS_PHRASES_BIO)
    jp_match = [k for k in ART_KEYWORDS_JP if k in bio_all]
    en_match = bool(ART_KEYWORDS_EN.search(bio_lower))
    site_match = [s for s in ART_SITES if s in site.lower()]

    if not (pos_phrase or jp_match or en_match or site_match):
        return False, "NO_SIGNAL"

    NON_ART = ["公式", "official", "情報発信", "グッズ", "フィギュア", "コスプレ", "cosplay", "ダンス", "dance", "singer", "vtuber", "料理", "レシピ", "グルメ"]
    neg_count = sum(1 for k in NON_ART if k.lower() in bio_lower)
    pos_count = (3 if pos_phrase else 0) + len(jp_match) + (1 if en_match else 0) + (2 if site_match else 0)

    if pos_count <= neg_count:
        return False, f"NEG_WIN pos={pos_count} neg={neg_count}"
    return True, f"pos={pos_count} neg={neg_count} jp={jp_match} en={en_match} site={site_match}"

def fetch_fx(entry):
    url = f"https://api.fxtwitter.com/{entry['screenName']}/status/{entry['id']}"
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "bot"})
        with urllib.request.urlopen(req, timeout=5) as r:
            return entry, json.loads(r.read()).get("tweet", {}).get("author")
    except:
        return entry, None

# Yahoo API取得
import urllib.parse
url = f"https://search.yahoo.co.jp/realtime/api/v1/pagination?p={urllib.parse.quote(QUERY)}&md=h&results={RESULTS}&mtype=image"
req = urllib.request.Request(url, headers=YAHOO_HEADERS)
with urllib.request.urlopen(req) as r:
    data = json.loads(r.read())

entries = [e for e in data["timeline"]["entry"] if e.get("media")]
print(f"クエリ: {QUERY}  メディアあり: {len(entries)}件\n")

# fxtwitter並列取得
with concurrent.futures.ThreadPoolExecutor(max_workers=10) as ex:
    fx_results = list(ex.map(fetch_fx, entries))

passed, rejected = [], []
for entry, author in fx_results:
    if not author:
        rejected.append((entry, author, "FX_FAIL"))
        continue
    ok, reason = is_illustrator(author)
    (passed if ok else rejected).append((entry, author, reason))

def show(entry, author, reason):
    name = author.get("name", "?") if author else "?"
    screen = entry["screenName"]
    desc = (author.get("description", "") or "")[:60].replace("\n", " ") if author else ""
    loc = (author.get("location", "") or "")[:30] if author else ""
    site = ((author.get("website", {}) or {}).get("url", "") or "")[:40] if author else ""
    text = (entry.get("displayTextBody", "") or "")[:60].replace("\n", " ")
    print(f"  @{screen} ({name})")
    print(f"    bio : {desc}")
    print(f"    loc : {loc}  site: {site}")
    print(f"    text: {text}")
    print(f"    → {reason}\n")

print(f"✅ 通過: {len(passed)}件")
for e, a, r in passed:
    show(e, a, r)

print(f"\n❌ 除外: {len(rejected)}件")
for e, a, r in rejected:
    show(e, a, r)

File: filter_debug.ts
#!/usr/bin/env tsx
/**
 * 使い方: npx tsx filter_debug.ts [クエリ] [件数=40]
 */
import { illustratorScore, isIllustrator, NEG_PHRASES_BIO, POS_PHRASES_BIO, ART_KEYWORDS_JP, ART_KEYWORDS_EN, ART_SITES } from "./src/lib/filter";

const QUERY = process.argv[2] ?? "イラスト";
const RESULTS = parseInt(process.argv[3] ?? "40");
const UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36";

function reason(author: any): string {
  const desc: string = author?.description ?? "";
  const loc: string = author?.location ?? "";
  const name: string = author?.name ?? "";
  const site: string = (author?.website?.url ?? "").toLowerCase();
  const bioAll = desc + " " + loc + " " + name;
  const bioAllLower = bioAll.toLowerCase();
  const allUrls = bioAllLower + " " + site;

  const negPhrase = NEG_PHRASES_BIO.find((p) => bioAll.includes(p));
  if (negPhrase) return `NEG_PHRASE: ${negPhrase}`;

  const posPhrase = POS_PHRASES_BIO.find((p) => bioAll.includes(p));
  const jpMatches = ART_KEYWORDS_JP.filter((k) => bioAll.includes(k));
  const enMatch = ART_KEYWORDS_EN.test(bioAllLower);
  const siteMatches = ART_SITES.filter((s) => allUrls.includes(s));

  if (!posPhrase && !jpMatches.length && !enMatch && !siteMatches.length) return "NO_SIGNAL";

  const NON_ART = ["公式", "official", "情報発信", "グッズ", "フィギュア", "コスプレ", "cosplay", "ダンス", "dance", "singer", "vtuber", "料理", "レシピ", "グルメ"];
  const negCount = NON_ART.filter((k) => bioAllLower.includes(k)).length;
  const posCount = (posPhrase ? 3 : 0) + jpMatches.length + (enMatch ? 1 : 0) + (siteMatches.length ? 2 : 0);

  if (posCount <= negCount) return `NEG_WIN pos=${posCount} neg=${negCount}`;
  return `pass pos=${posCount} neg=${negCount} jp=${JSON.stringify(jpMatches)} en=${enMatch} site=${JSON.stringify(siteMatches)}${posPhrase ? ` phrase="${posPhrase}"` : ""}`;
}

async function main() {
  const yahooUrl = `https://search.yahoo.co.jp/realtime/api/v1/pagination?p=${encodeURIComponent(QUERY)}&md=h&results=${RESULTS}&mtype=image`;
  const data = await fetch(yahooUrl, {
    headers: { "User-Agent": UA, "Accept": "application/json", "Referer": "https://search.yahoo.co.jp/realtime/search" },
  }).then((r) => r.json());

  const entries: any[] = (data?.timeline?.entry ?? []).filter((e: any) => e.media?.length > 0);
  console.log(`クエリ: ${QUERY}  メディアあり: ${entries.length}件\n`);

  const fxResults = await Promise.all(
    entries.map((e) =>
      fetch(`https://api.fxtwitter.com/${e.screenName}/status/${e.id}`, { headers: { "User-Agent": "bot" } })
        .then((r) => r.ok ? r.json() : null).catch(() => null)
    )
  );

  const passed: any[] = [], rejected: any[] = [];
  for (let i = 0; i < entries.length; i++) {
    const e = entries[i];
    const author = fxResults[i]?.tweet?.author;
    const r = reason(author);
    const ok = author && isIllustrator(author, e.displayTextBody ?? "");
    (ok ? passed : rejected).push({ e, author, r });
  }

  function show({ e, author, r }: any) {
    const desc = (author?.description ?? "").slice(0, 60).replace(/\n/g, " ");
    const loc = (author?.location ?? "").slice(0, 30);
    const site = (author?.website?.url ?? "").slice(0, 40);
    const bioUrls = (author?.description ?? "").match(/https?:\/\/\S+/g)?.join(" ").slice(0, 60) ?? "";
    const text = (e.displayTextBody ?? "").replace(/\tSTART\t|\tEND\t/g, "").slice(0, 60).replace(/\n/g, " ");
    const score = author ? illustratorScore(author, e.hashtags ?? [], e.displayTextBody ?? "") : 0;
    console.log(`  @${e.screenName} (${author?.name ?? "?"}) score=${score}`);
    console.log(`    bio : ${desc}`);
    console.log(`    loc : ${loc}`);
    if (site) console.log(`    site: ${site}`);
    if (bioUrls) console.log(`    urls: ${bioUrls}`);
    console.log(`    text: ${text}`);
    console.log(`    → ${r}\n`);
  }

  console.log(`✅ 通過: ${passed.length}件`);
  passed.forEach(show);
  console.log(`\n❌ 除外: ${rejected.length}件`);
  rejected.forEach(show);
}

main();
